libraries/AP_Airspeed/AP_Airspeed_ASP5033.cpp

```diff
@@ -0,0 +1,175 @@
+        checkwhoami != REG_WHOAMI_DEFAULT_ID){
+        _dev->set_retries(2);
+        return false;      
+    }
+
+    //Confirm it's ASP5033 sensor, not the other.
+    bool confirm = false;
+    confirm = Sensor_confirm();
+    if (!confirm) {
+        return false;
+    }
+
+    WITH_SEMAPHORE(sem);
+    uint8_t SysID;       
+    if (_dev->read_registers(REG_SYS_CONFIG, &SysID, 1)|| //Read Sensor_config
+        SysID != (SysID | RAW_DATA_ON)){ //Read the 0xA5 register value, put the read binary value "and" on "11111101" then write to 0xA5.
+    _dev->write_register(REG_SYS_CONFIG, SysID);     //Set Sensor_config
+    }
+    _dev->register_periodic_callback(1000000UL/80U,
+                                    FUNCTOR_BIND_MEMBER(&AP_Airspeed_ASP5033::timer, void));
+    return true;
+}
+
+bool AP_Airspeed_ASP5033::Sensor_confirm(void)
+{
+    //Confirm it's ASP5033 sensor, not the other.
+    //Set Whoami ID. 0XA4 is a OPT WhoamI Part ID register, also can be read from address 0x01.
+    //check whoami, set the PartID register to 0X66 first, them Read the PartID register.
+    uint8_t whoami;
+    _dev->write_register(REG_WAI_SET, REG_WHOAMI_RECHECK_ID);
+    if (!_dev->read_registers(REG_WAI, &whoami, 1) || //Read the whoami ID
+        whoami != REG_WHOAMI_RECHECK_ID){
+        return false; // it maybe not a ASP5033, tuneoff ASP5033
+    }   
+    else{
+        return true;  
+    }    
+}
+
+
+// read the data from the sensor
+void AP_Airspeed_ASP5033::timer()
+{
+    _dev->write_register(REG_CMD, CMD_MEASURE); //once temperature conversion immediately followed by once sensor signal conversion
+    uint8_t SoC; //Start of conversion flag
+    bool sensing = true;    //sensor m
+    //uint8_t Ready = 0x01;
+    if (_dev->read_registers(REG_CMD, &SoC, 1)){ //Judge whether Data collection is over
+    if ((SoC & REG_SENSOR_READY) != sensing){   // Read the 0x30 register address. If Sco bit is 0,signify the acquisition end, the data can be read.
+        uint8_t pressraw_bytes[2]; //pressure data AD value       
+        if(_dev->read_registers(REG_PRESS_DATA, pressraw_bytes, 2)){
+        //Read ADC output data of pressure
+        //Read 0x06, 0x07, 0x08 register address data to form a 24-bit AD value (pressure data AD value).
+        _pressraw = (pressraw_bytes[0]<<24) + (pressraw_bytes[1]<<16); 
+        //Effective value converting 24 bit integer to 32 bit integer 
+        _pressraw >>=15;
+        }        
+        //Read ADC output data of temperature
+        //Read 0x09, 0x10, register address data to form a 16-bit AD value (temperature data AD value,the highest is the symbol bit.).
+        uint8_t tempraw_bytes[2]; //temperature data AD value
+        if (_dev->read_registers(REG_TEMP_DATA, tempraw_bytes, 1)){
+        _tempraw = tempraw_bytes[0];
+        //Right-shift signed integer back to get correct measurement value
+        }  
+
+        WITH_SEMAPHORE(sem); 
+        _press_sum += _pressraw;
```

Comment:
You have to handle the incrementing seprately for the pressure and the temperature. You can only increment the average if you got a new value off the bus (IE the read above worked, if the read didn't work then you can't add it to the average. If you fix the _tempraw being a member of the class this will become an obvious fix).